### PR TITLE
implement findAllByGroup

### PR DESCRIPTION
**Pull request by frett**
I needed the functionality to list out all the members in a specific Group. In the process I also cleaned up a few corner case bugs with group parsing.

### CODE REVIEW

**Comment by frett** [comment]
@hlbraddock

**Comment by frett** [comment]
group testing properties for our test eDirectory cluster:

ldap.dn.group=ou=grouptest,dc=ccci,dc=org
ldap.dn.group.group1=cn=group1,ou=grouptest,dc=ccci,dc=org
ldap.dn.group.group2=cn=group2,ou=subtree,ou=grouptest,dc=ccci,dc=org

File: ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java [hunk 1]
```diff
@@ -41,6 +45,13 @@
     @Inject
     private LdaptiveUserDao dao;
 
+    @Inject
```
Comment (hlbraddock) [inline ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java]: :+1: This is a nice feature.

File: ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java [hunk 2]
```diff
@@ -364,78 +383,100 @@ public void testFindUserByEmployeeId() throws Exception {
...
-        String name = "Mail";
-
-        List<Group> groups = Lists.newArrayList();
-        for(String[] path : paths) {
-            Group group = new Group(path, name);
-            groups.add(group);
-            this.dao.addToGroup(user, group);
+        // assert the 2 users are not in the group
+        {
+            final Set<String> guids = FluentIterable.from(this.dao.findAllByGroup(group1, true)).transform
```
Comment (hlbraddock) [inline ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java]: Do you find it easier to use `FluentIterable` with `Function` converting `User` to `Set<String>` in order to check with `contains`? Or for some other reason?

Comment (frett) [inline ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java]: because the `User` object is different after adding them to a `Group`, `contains(user)` doesn't work

File: ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java [hunk 3]
```diff
@@ -364,78 +383,100 @@ public void testFindUserByEmployeeId() throws Exception {
...
+        {
+            this.dao.removeFromGroup(user, group2);
+            final User foundUser = this.dao.findByEmail(user.getEmail(), false);
+            assertEquals(ImmutableSet.of(), foundUser.getGroups());
         }
     }
 
     @Test
-    public void testRemoveFromGroup() throws Exception {
+    public void testFindAllByGroup() throws Exception {
```
Comment (hlbraddock) [inline ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java]: +1 `testFindAllByGroup` seems to thoroughly test API.

**Comment by hlbraddock** [comment]
+1 This PR looks good. I didn't run the tests, but I'm confident you did and that they passed.

File: ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java [hunk 4]
```diff
@@ -364,78 +383,100 @@ public void testFindUserByEmployeeId() throws Exception {
...
-        final User user = getUser();
+        // create a couple users for testing
+        final User user1 = getUser();
+        final User user2 = getUser();
+        user2.setDeactivated(true);
+        this.dao.save(user1);
+        this.dao.save(user2);
 
-        this.dao.save(user);
+        final Function<User, String> FUNCTION_GUID = new Function<User, String>() {
```
Comment (frett) [inline ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java]: @hlbraddock any opinions on moving this function to the `User` object so it can potentially be used by projects utilizing this library to transform lists of users?

I know in The Key I've implemented a `Group.getName` function that gets used in a couple places that I've thought about moving to the `Group` object.

File: ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java (hunk 2 above)
Comment (hlbraddock) [inline ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java]: +1

File: ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java (hunk 4 above)
Comment (hlbraddock) [inline ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java]: My two cents.

There always seems to be tradeoff between providing certain potentially valuable (but currently unused) functionality (e.g. `IllegalGroupDnException`) and/or 'creeping elegance' in the code vs. leaving it undeveloped until requirements warrant it. I don't have a hard opinion either way. I would simply say that if you've felt the need to develop the functionality or add the improvement, then it's probably worth providing it for potential future use.

So, I would say add `Function` `FUNCTION_GUID` to the `User` object, the only caveat being that there be clear delineation between types of methods, that is, getter/setters vs not meant for public use (e.g. `public <T> T getImplMeta`), and this helper method. Whether you distinguish by way of comments and method placement in the class, or whether you provide another 'helper' class, I leave to you, though I would lean towards the former, as you originally suggested.

Comment (frett) [inline ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java]: I was more suggesting providing this function object as a static constant on the user object (e.g. `User.FUNCTION_GUID`). So it wouldn't be any new methods, it would just be a guava function object that could be referenced directly instead of needing to be duplicated everywhere it's needed. I'm going to leave this here until it's needed elsewhere, then move it at that point to prevent code duplication.

Comment (hlbraddock) [inline ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoIT.java]: +1
